app/assets/javascripts/validation.js

```diff
@@ -10,69 +10,228 @@ $(document).ready(function () {
+  var emailElement = document.querySelector(
+    formClass + " [name='user[email]']"
+  );
+  var passwordElement = document.querySelector(
+    formClass + " [name='user[password]']"
+  );
+  var confirmPasswordElement = document.querySelector(
+    formClass + " [name='user[password_confirmation]']"
+  );
+
+  // Every time user types something, it triggers corresponding event listener
+  usernameElement.addEventListener(
+    'input',
+    validateUsername.bind(usernameElement, this)
+  );
+  emailElement.addEventListener(
+    'input',
+    validateEmail.bind(emailElement, this)
+  );
+  passwordElement.addEventListener(
+    'input',
+    validatePassword.bind(passwordElement, confirmPasswordElement, this)
+  );
+  confirmPasswordElement.addEventListener(
+    'input',
+    validateConfirmPassword.bind(confirmPasswordElement, passwordElement, this)
+  );
+}
+
+// Every time user types something in the form, it triggers this function
+// It will update UI depending on the value of <valid> parameter
+SignUpFormValidator.prototype.updateUI = function(element, valid, errorMsg) {
+  var elementName = element.getAttribute('name');
+
+  if (valid) {
+    this.validationTracker[elementName] = true;
+    styleElement(element, 'form-element-invalid', 'form-element-valid');
+    removeErrorMsg(element);
+  } else {
+    this.validationTracker[elementName] = false;
+    styleElement(element, 'form-element-valid', 'form-element-invalid');
+    renderErrorMsg(element, errorMsg);
+  }
+
+  this.isFormValid();
+};
+
+SignUpFormValidator.prototype.disableSubmitBtn = function() {
+  this.submitBtn.setAttribute('disabled', '');
+};
+
+SignUpFormValidator.prototype.enableSubmitBtn = function() {
+  this.submitBtn.removeAttribute('disabled');
+};
+
+SignUpFormValidator.prototype.isFormValid = function() {
+  // Form is valid if all elements have passsed validation successfully
+  var isValid =
+    Object.values(this.validationTracker).filter(Boolean).length === 4;
+
+  if (isValid) this.enableSubmitBtn();
+  else this.disableSubmitBtn();
+};
+
+function validateUsername(obj) {
+  var username = this.value;
+  var self = this;
+
+  if (username.length < 3) {
+    restoreOriginalStyle(this);
+  } else {
+    $.get('/api/srch/profiles?query=' + username, function(data) {
+      if (data.items) {
+        $.map(data.items, function(userData) {
+          if (userData.doc_title === username) {
+            obj.updateUI(self, false, 'Username already exists');
+          } else {
+            obj.updateUI(self, true);
+          }
+        });
+      } else {
+        obj.updateUI(self, true);
       }
-    },
-    submitHandler: function(form) {
-      form.submit();
-    }
-  });
+    });
+  }
+}
+
+function validateEmail(obj) {
+  var email = this.value;
+  var regexp = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+  var isValid = regexp.test(email);
```

Comment:
What is valid?

Reply:
isValid is a boolean value that is returned from the regular expression.
It determines if the email is valid or not.

Reply:
That is common practice to prefix variables that contain a Boolean value(false/true) with "is" or "has".

Reply:
I know what it means. I meant it should be isValidEmail.

Reply:
Alright, I'll update it.